Firestore/core/test/firebase/firestore/testutil/async_testing.cc

```diff
@@ -36,6 +40,62 @@ std::shared_ptr<util::AsyncQueue> AsyncQueueForTesting() {
   return std::make_shared<AsyncQueue>(ExecutorForTesting("worker"));
 }
 
+// MARK: - Expectation
+
+Expectation::Expectation() : promise_(std::make_shared<std::promise<void>>()) {
+}
+
+void Expectation::Fulfill() {
+  promise_->set_value();
+}
+
+std::function<void()> Expectation::AsCallback() const {
+  // Additional copy required because putting `promise_` in the capture list
+  // is not allowed. Generalized capture in C++14 would allow direct
+  // initialization.
+  auto promise = promise_;
+  return [promise] { promise->set_value(); };
+}
+
+std::future<void> Expectation::get_future() const {
+  return promise_->get_future();
+}
+
+// MARK: - AsyncTest
+
+std::future<void> AsyncTest::Async(std::function<void()> action) {
+  std::packaged_task<void()> task(std::move(action));
+  auto future = task.get_future();
+
+  std::thread thread(std::move(task));
+  thread.detach();
+  return future;
+}
+
+void AsyncTest::Await(const std::future<void>& future,
+                      const std::chrono::milliseconds timeout) {
```

Comment:
Optional: remove `const` from `timeout` (I'm all for having local variables, but I think we have previously agreed to not have them, so this is inconsistent with the rest of the codebase). Applies below as well.